water-effect/particle.js

```diff
@@ -0,0 +1,244 @@
+      return;
+    }
+    const timestamp = performance.now();
+
+    //#################################### handle diving water ####################################
+    if (this.contactWater) {
+      this.lastContactWaterTime = timestamp;
+    }
+    if (this.contactWater && this.lastContactWater !== this.contactWater) {
+      this.fallingSpeed = 0 - this.player.characterPhysics.velocity.y;
+      this.divingCollisionPos.set(this.player.position.x, this.waterSurfaceHeight, this.player.position.z);
+
+      // play the diving particle
+      if (this.fallingSpeed > 1) {
+        this.playDivingSfx();
+        this.rippleMesh && this.emitDivingRipple();
+        this.divingLowerSplash && this.emitDivingLowerSplash();
+        this.divingHigherSplash && this.emitDivingHigherSplash();
+      }
+    }
+    else {
+      this.fallingSpeed = 0;
+    }
+   
+    // update particle
+    this.rippleMesh.update();
+    this.divingLowerSplash.update();
+    this.divingHigherSplash.update();
+
+    this.lastContactWater = this.contactWater;
+    this.scene.updateMatrixWorld();
+  }
+
+  //########################################################## initialize particle mesh #####################################################
+  initRipple() {
+    this.rippleGroup = getDivingRipple(this.models.ripple);
+    this.rippleMesh = this.rippleGroup.children[0].children[0];
+    this.rippleMesh.material.uniforms.rippleTexture.value = this.textures.rippleTexture;
+    this.rippleMesh.material.uniforms.voronoiNoiseTexture.value = this.textures.voronoiNoiseTexture;
+    this.rippleMesh.material.uniforms.noiseMap.value = this.textures.noiseMap;
+    this.rippleMesh.update = () => this.updateDivingRipple();
+    this.scene.add(this.rippleGroup);
+  }
+  initDivingLowerSplash() {
+    this.divingLowerSplash = getDivingLowerSplash();
+    this.divingLowerSplash.material.uniforms.splashTexture.value = this.textures.splashTexture2;
+    this.divingLowerSplash.material.uniforms.noiseMap.value = this.textures.noiseMap;
+    this.divingLowerSplash.update = () => this.updateDivingLowerSplash();
+    this.scene.add(this.divingLowerSplash);
+  }
+  initDivingHigherSplash() {
+    this.divingHigherSplash = getDivingHigherSplash();
+    this.divingHigherSplash.material.uniforms.splashTexture.value = this.textures.splashTexture3;
+    this.divingHigherSplash.material.uniforms.noiseMap.value = this.textures.noiseMap;
+    this.divingHigherSplash.update = () => this.updateDivingHigherSplash();
+    this.scene.add(this.divingHigherSplash);
+  }
+
+  //########################################################## update particle mesh #####################################################
+  updateDivingRipple() {
+    if (this.rippleMesh) {
+      const falling = this.fallingSpeed > 10 ? 10 : this.fallingSpeed;
+      if (this.rippleMesh.material.uniforms.vBroken.value < 1) {
+        if (this.rippleMesh.scale.x > 0.15 * (1 + falling * 0.1)) {
```

Comment:
separate this math, possibly try to explain all of the math you're doing

Reply:
[use constant](https://github.com/webaverse/heightfield/pull/13/commits/7f81692533d3a8bad44a7572c28bd7b5a5025042)